### DISCUSSION TITLE
If vs Else If performance test - Cairo

### DISCUSSION THREAD

**Discussion by danilowhk** [kkrt-labs/kakarot-ssj#77] [General]
# Comparing Multiple Consecutive `if` vs `if-else` Statements: A Benchmark

## Introduction
This report compares the gas cost of using multiple consecutive `if` statements versus using `if-else` statements. The tests were conducted considering 20 conditions, utilizing scarb 0.6.0-alpha.2 and cairo 2.1.0-rc2.

## Methodology
The benchmarks were conducted under the following configurations:
- **Language:** Cairo
- **Version:** scarb 0.6.0-alpha.2, cairo 2.1.0-rc2
- **Conditions:** 20 different conditions were evaluated
- **Metrics:** Gas consumption

## Overall Result
The comparison between consecutive 20 consecutive if statements and 20 consecutive if-else statements highlight a significant difference in gas consumption:
```
| Approach                | Gas Consumption |
|-------------------------|-----------------|
| Consecutive `if`        | 7620           |
| Consecutive `if-else`   | 7620           |
```

### Description
This approach uses consecutive `if` statements, where each condition is evaluated independently.

### Code
```cairo
fn test_if_function(a: felt252) -> felt252{
    if a == 1 {
        return a;
    }
    if a == 2 {
        return a;
    }
    if a == 3 {
        return a;
    }
    if a == 4 {
        return a;
    }
    if a == 5 {
        return a;
    }
    if a == 6 {
        return a;
    }
    if a == 7 {
        return a;
    }
    if a == 8 {
        return a;
    }
    if a == 9 {
        return a;
    }
    if a == 10 {
        return a;
    }

    if a == 11 {
        return a;
    }
    if a == 12 {
        return a;
    }
    if a == 13 {
        return a;
    }
    if a == 14 {
        return a;
    }
    if a == 15 {
        return a;
    }
    if a == 16 {
        return a;
    }
    if a == 17 {
        return a;
    }
    if a == 18 {
        return a;
    }
    if a == 19 {
        return a;
    }
    if a == 20 {
        return a;
    }
    return a;

}


#[test]
#[available_gas(1000000)]
fn test_if() {
    let x = testing::get_available_gas();
    gas::withdraw_gas().unwrap();
    let a = 20;
    let value = test_if_function(a);

    (x - testing::get_available_gas()).print();
}
```
### Result
Gas consumption for this test was 7620.

### Description
This approach uses if-else statements, evaluating conditions sequentially and exiting when a true condition is met.

### Code
```cairo
fn test_if_else_function(a: felt252) -> felt252 {

    if a == 1 {
        return a;
    } else if a == 2 {
        return a;
    } else if a == 3 {
        return a;
    } else if a == 4 {
        return a;
    } else if a == 5 {
        return a;
    } else if a == 6 {
        return a;
    } else if a == 7 {
        return a;
    } else if a == 8 {
        return a;
    } else if a == 9 {
        return a;
    } else if a == 10 {
        return a;
    } else if a == 11 {
        return a;
    } else if a == 12 {
        return a;
    } else if a == 13 {
        return a;
    } else if a == 14 {
        return a;
    } else if a == 15 {
        return a;
    } else if a == 16 {
        return a;
    } else if a == 17 {
        return a;
    } else if a == 18 {
        return a;
    } else if a == 19 {
        return a;
    } else if a == 20 {
        return a;
    }
    return a;

}

#[test]
#[available_gas(1000000)]
fn test_if_else() {
    let x = testing::get_available_gas();

    gas::withdraw_gas().unwrap();
    let a = 20;
    let value = test_if_else_function(a);

    (x - testing::get_available_gas()).print();
}

```

### Result
Gas consumption for this test was 7620.

## Conclusion
The tests indicate that using if-else costs the same amount of steps compared to consecutive ifs.

### Versions used:
scarb: 0.6.0-alpha.2 (4b4d5411b 2023-07-25)
cairo: 2.1.0-rc2 (https://crates.io/crates/cairo-lang-compiler/2.1.0-rc2)

**Comment by enitrat** [comment]
would that make a difference if you had return statements in each branch? Here, you're not really comparing two similar behaviors.
